src/sinks/util/buffer/compression.rs

```diff
@@ -1,20 +1,52 @@
 use std::fmt;
 
 use serde::{de, ser};
-use serde_json::Value;
+use vector_config::{
+    configurable_component,
+    schema::{
+        finalize_schema, generate_composite_schema, generate_number_schema, generate_string_schema,
+    },
+    schemars::{gen::SchemaGenerator, schema::SchemaObject},
+    validation::Validation,
+    Configurable, Metadata,
+};
 
-pub const GZIP_NONE: u32 = 0;
-pub const GZIP_FAST: u32 = 1;
-pub const GZIP_DEFAULT: u32 = 6;
-pub const GZIP_BEST: u32 = 9;
+// NOTE: The resulting schema for `Compression` is not going to look right, because of how we do the customized "string
+// or map" stuff. We've done a bunch of work here to get us half-way towards generating a valid schema for it, including
+// add in the `CompressionLevel` type and doing a custom `Configurable` implementation for it... but the real meat of
+// the change would be doing so for `Compression` because of it being an enum and all the boilerplate involved there.
+//
+// It'd be nice to find a succinct way to allow the expression of these constraints, possibly through a declarative
+// macro that helps build the manual `Configurable` implementation such that, while still leaving the developer
+// responsible for generating a correct schema, it would be far closer to a human-readable description of the schema
+// i.e. "this schema can be X or Y, and if X, it can only be these values, or if Y, the value must fall within this
+// specific range" and so on, with methods for describing enum variants, etc.
+//
+// Another complication is that we can _almost_ sort of get there with the existing primitives but `serde` itself has no
+// way to express an enum that functions like what we're doing here i.e. the variant name is parsed from the value,
+// without needing to be a tag field, with the ability to fallback to a newtype variant, etc.
+//
+// Long story short, these custom (de)serialization strategies are hard to encode in a procedural way because they are
+// in fact not themselves declared procedurally.
```

Comment:
Isn't the problem that the custom implementations are declared procedurally rather than declaratively?